File: src/color_science/transforms.ts
// Konversi Warna Presisi: sRGB <-> Linear RGB <-> CIE XYZ (D65) <-> CIE L*a*b*

export interface RGB {
  r: number; // 0 - 255
  g: number; // 0 - 255
  b: number; // 0 - 255
}

export interface XYZ {
  x: number; // 0 - 100+
  y: number; // 0 - 100
  z: number; // 0 - 100+
}

export interface Lab {
  l: number; // 0 - 100 (Kecerahan / Lightness)
  a: number; // -128 - +127 (Hijau - Merah)
  b: number; // -128 - +127 (Biru - Kuning)
}

// Standar Iluminan D65 (2 derajat pengamat standar CIE 1931)
const D65_XN = 95.047;
const D65_YN = 100.000;
const D65_ZN = 108.883;

/**
 * Konversi sRGB (0-255) ke Linear RGB (0-1)
 */
export function sRGBToLinearRGB(rgb: RGB): { r: number; g: number; b: number } {
  const toLinear = (c: number) => {
    const v = c / 255;
    return v <= 0.04045 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
  };
  return {
    r: toLinear(rgb.r),
    g: toLinear(rgb.g),
    b: toLinear(rgb.b),
  };
}

/**
 * Konversi Linear RGB (0-1) ke sRGB (0-255)
 */
export function linearRGBToSRGB(linear: { r: number; g: number; b: number }): RGB {
  const toSRGB = (v: number) => {
    const clamped = Math.max(0, Math.min(1, v));
    const c = clamped <= 0.0031308 ? 12.92 * clamped : 1.055 * Math.pow(clamped, 1 / 2.4) - 0.055;
    return Math.round(c * 255);
  };
  return {
    r: toSRGB(linear.r),
    g: toSRGB(linear.g),
    b: toSRGB(linear.b),
  };
}

/**
 * Konversi sRGB ke CIE XYZ
 */
export function rgbToXyz(rgb: RGB): XYZ {
  const lin = sRGBToLinearRGB(rgb);
  // Matriks transformasi sRGB ke XYZ (D65)
  const x = (lin.r * 0.4124564 + lin.g * 0.3575761 + lin.b * 0.1804375) * 100;
  const y = (lin.r * 0.2126729 + lin.g * 0.7151522 + lin.b * 0.0721750) * 100;
  const z = (lin.r * 0.0193339 + lin.g * 0.1191920 + lin.b * 0.9503041) * 100;
  return { x, y, z };
}

/**
 * Konversi CIE XYZ ke CIE L*a*b*
 */
export function xyzToLab(xyz: XYZ): Lab {
  const fx = fXYZ(xyz.x / D65_XN);
  const fy = fXYZ(xyz.y / D65_YN);
  const fz = fXYZ(xyz.z / D65_ZN);

  const l = Math.max(0, 116 * fy - 16);
  const a = 500 * (fx - fy);
  const b = 200 * (fy - fz);

  return { l, a, b };
}

function fXYZ(t: number): number {
  const delta = 6 / 29;
  return t > Math.pow(delta, 3) ? Math.cbrt(t) : t / (3 * Math.pow(delta, 2)) + 4 / 29;
}

/**
 * Konversi langsung sRGB ke CIE L*a*b*
 */
export function rgbToLab(rgb: RGB): Lab {
  return xyzToLab(rgbToXyz(rgb));
}

/**
 * Konversi CIE L*a*b* ke CIE XYZ
 */
export function labToXyz(lab: Lab): XYZ {
  const fy = (lab.l + 16) / 116;
  const fx = lab.a / 500 + fy;
  const fz = fy - lab.b / 200;

  const invF = (t: number) => {
    const delta = 6 / 29;
    return t > delta ? Math.pow(t, 3) : 3 * Math.pow(delta, 2) * (t - 4 / 29);
  };

  const x = D65_XN * invF(fx);
  const y = D65_YN * invF(fy);
  const z = D65_ZN * invF(fz);

  return { x, y, z };
}

/**
 * Konversi CIE XYZ ke sRGB
 */
export function xyzToRgb(xyz: XYZ): RGB {
  const x = xyz.x / 100;
  const y = xyz.y / 100;
  const z = xyz.z / 100;

  const r = x *  3.2404542 + y * -1.5371385 + z * -0.4985314;
  const g = x * -0.9692660 + y *  1.8760108 + z *  0.0415560;
  const b = x *  0.0556434 + y * -0.2040259 + z *  1.0572252;

  return linearRGBToSRGB({ r, g, b });
}

/**
 * Konversi CIE L*a*b* kembali ke sRGB
 */
export function labToRgb(lab: Lab): RGB {
  return xyzToRgb(labToXyz(lab));
}

File: src-tauri/src/color_science/transforms.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct Rgb {
    pub r: u8,
    pub g: u8,
    pub b: u8,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct Xyz {
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct Lab {
    pub l: f64, // 0.0 - 100.0
    pub a: f64, // -128.0 to +127.0
    pub b: f64, // -128.0 to +127.0
}

// Standar Iluminan D65 (2 derajat pengamat standar CIE 1931)
const D65_XN: f64 = 95.047;
const D65_YN: f64 = 100.000;
const D65_ZN: f64 = 108.883;

pub fn srgb_to_linear(c: u8) -> f64 {
    let v = c as f64 / 255.0;
    if v <= 0.04045 {
        v / 12.92
    } else {
        ((v + 0.055) / 1.055).powf(2.4)
    }
}

pub fn linear_to_srgb(v: f64) -> u8 {
    let clamped = v.clamp(0.0, 1.0);
    let c = if clamped <= 0.0031308 {
        12.92 * clamped
    } else {
        1.055 * clamped.powf(1.0 / 2.4) - 0.055
    };
    (c * 255.0).round() as u8
}

pub fn rgb_to_xyz(rgb: &Rgb) -> Xyz {
    let r_lin = srgb_to_linear(rgb.r);
    let g_lin = srgb_to_linear(rgb.g);
    let b_lin = srgb_to_linear(rgb.b);

    let x = (r_lin * 0.4124564 + g_lin * 0.3575761 + b_lin * 0.1804375) * 100.0;
    let y = (r_lin * 0.2126729 + g_lin * 0.7151522 + b_lin * 0.0721750) * 100.0;
    let z = (r_lin * 0.0193339 + g_lin * 0.1191920 + b_lin * 0.9503041) * 100.0;

    Xyz { x, y, z }
}

fn f_xyz(t: f64) -> f64 {
    let delta: f64 = 6.0 / 29.0;
    if t > delta.powi(3) {
        t.cbrt()
    } else {
        t / (3.0 * delta.powi(2)) + 4.0 / 29.0
    }
}

pub fn xyz_to_lab(xyz: &Xyz) -> Lab {
    let fx = f_xyz(xyz.x / D65_XN);
    let fy = f_xyz(xyz.y / D65_YN);
    let fz = f_xyz(xyz.z / D65_ZN);

    let l = (116.0 * fy - 16.0).max(0.0);
    let a = 500.0 * (fx - fy);
    let b = 200.0 * (fy - fz);

    Lab { l, a, b }
}

pub fn rgb_to_lab(rgb: &Rgb) -> Lab {
    xyz_to_lab(&rgb_to_xyz(rgb))
}

pub fn lab_to_xyz(lab: &Lab) -> Xyz {
    let fy = (lab.l + 16.0) / 116.0;
    let fx = lab.a / 500.0 + fy;
    let fz = fy - lab.b / 200.0;

    let inv_f = |t: f64| -> f64 {
        let delta: f64 = 6.0 / 29.0;
        if t > delta {
            t.powi(3)
        } else {
            3.0 * delta.powi(2) * (t - 4.0 / 29.0)
        }
    };

    let x = D65_XN * inv_f(fx);
    let y = D65_YN * inv_f(fy);
    let z = D65_ZN * inv_f(fz);

    Xyz { x, y, z }
}

pub fn xyz_to_rgb(xyz: &Xyz) -> Rgb {
    let x = xyz.x / 100.0;
    let y = xyz.y / 100.0;
    let z = xyz.z / 100.0;

    let r_lin = x * 3.2404542 + y * -1.5371385 + z * -0.4985314;
    let g_lin = x * -0.9692660 + y * 1.8760108 + z * 0.0415560;
    let b_lin = x * 0.0556434 + y * -0.2040259 + z * 1.0572252;

    Rgb {
        r: linear_to_srgb(r_lin),
        g: linear_to_srgb(g_lin),
        b: linear_to_srgb(b_lin),
    }
}

pub fn lab_to_rgb(lab: &Lab) -> Rgb {
    xyz_to_rgb(&lab_to_xyz(lab))
}
